src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -98,16 +99,26 @@ bool BoundsCheckInfo::Initialize(const Compiler* comp, Statement* statement, Gen
 // RemoveBoundsChk - Remove the given bounds check from the statement and the block.
 //
 // Arguments:
-//    comp  - compiler instance
-//    check - bounds check node to remove
-//    comma - check's parent node (either null or COMMA)
-//    stmt  - statement containing the bounds check
+//    comp    - compiler instance
+//    treeUse - the bounds check node to remove (its use edge)
+//    stmt    - the statement containing the bounds check
 //
 static void RemoveBoundsChk(Compiler* comp, GenTree** treeUse, Statement* stmt)
 {
     JITDUMP("Before RemoveBoundsChk:\n");
     DISPTREE(*treeUse);
 
+    // If we deal with GT_ARR_LEN(lcl), we can unmark the array length node
+    // as having an NRE. Technically, this should be done by the assertion prop,
+    // but not today, see https://github.com/dotnet/runtime/pull/93531
+    //
+    GenTreeBoundsChk* bndsChk = (*treeUse)->AsBoundsChk();
+    GenTree*          arrObj  = bndsChk->GetArray();
+    if ((arrObj != nullptr) && ((arrObj->gtFlags & GTF_ALL_EFFECT) == 0))
+    {
+        bndsChk->GetArrayLength()->gtFlags &= ~GTF_EXCEPT;
+    }
```

Comment:
This isn't really correct. The only thing we know here is that the conservative normal VN is the same as the first bounds check's conservative normal VN, but the array length here could still be something like `0 * (1 / 0) + arr.Length` (obviously a constructed example) where this code would get rid of the `DivisionByZeroException`.

Maybe something like
```cpp
if (bndsChk->GetArrayLength()->OperIs(GT_ARR_LENGTH) &&
    firstBndsChk->GetArrayLength()->OperIs(GT_ARR_LENGTH) &&
    (bndsChk->GetArrayLength()->GetIndirOrArrMetaDataAddr()->gtVNPair.GetConservative() == firstBndsChk->GetArrayLength()->GetIndirOrArrMetaDataAddr()->gtVNPair.GetConservative())
{
  bndsChk->GetArrayLength()->gtFlags |= GTF_IND_NONFAULTING;
  gtUpdateStmtSideEffects(stmt);
}
```
would get the cases?

Reply:
Actually, even previous code didn't catch anything, so let me just remove this part